### PR TITLE
[Fix] Multi NIC IP Alignment

### PR DESCRIPTION
**Pull request by Nbot07A**
### Description of changes
* set ip to NIC 0 Device 0 primary ip to ensure IP alignment with Multi NIC instances

### Tests
* Passed new unit test test_from_describe_instance_data 
* Added multi-nic structure to test instances

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/slurm_plugin/fleet_manager.py [hunk 1]
```diff
@@ -47,12 +47,15 @@ def __hash__(self):
...
-            )
+            for network_interface in instance_info["NetworkInterfaces"]:
+                attachment = network_interface["Attachment"]
+                if attachment["DeviceIndex"] == 0 and attachment["NetworkCardIndex"] == 0:
+                    return EC2Instance(
+                        instance_info["InstanceId"],
+                        network_interface["PrivateIpAddress"],
+                        instance_info["PrivateDnsName"].split(".")[0],
+                        instance_info["LaunchTime"],
+                    )
```
Comment (francesco-giordano) [inline src/slurm_plugin/fleet_manager.py]: **Suggestion:** Can you refactor the code in order to address CodeQL issue.
An approach could be assign `instance_info["PrivateIpAddress"]` as a starting value to a variable. In the loop try to assign the right `network_interface["PrivateIpAddress"]` to that variable. Then use the variable to set the second field of the `return EC2Instance`.
In this way  there is the fallback of `instance_info["PrivateIpAddress"]` and a single explicit return.

**Review by francesco-giordano — COMMENTED** [review]
Have all the unit test passed?

**Comment by Nbot07A** [comment]
> Have all the unit test passed?

Yes

**Review by francesco-giordano — APPROVED** [review]
LGTM!

**Review by chenwany — COMMENTED** [review]
We need to make sure the private ip here is the one return from ec2 describe-instances in multi-nic instances

**Review by chenwany — CHANGES_REQUESTED** [review]
We need to update describe_instances together in the PR, or we will introduce another problem: during the iteration of clustermgtd, the multi-nic slurm node will be terminated due to ip mismatch.
We need to update here accordingly:
https://github.com/aws/aws-parallelcluster-node/blob/3d69d7ad5c1b43308d598a9433c820a8d800d0eb/src/slurm_plugin/clustermgtd.py#L1028

**Comment by gmarciani** [comment]
To be done in this PR: we must align the PrivateIpAddress also here

https://github.com/aws/aws-parallelcluster-node/blob/3d69d7ad5c1b43308d598a9433c820a8d800d0eb/src/slurm_plugin/instance_manager.py#L345

File: src/slurm_plugin/fleet_manager.py [hunk 2]
```diff
@@ -47,9 +47,14 @@ def __hash__(self):
     @staticmethod
     def from_describe_instance_data(instance_info):
         try:
+            private_ip = instance_info["PrivateIpAddress"]
+            for network_interface in instance_info["NetworkInterfaces"]:
+                attachment = network_interface["Attachment"]
+                if attachment["DeviceIndex"] == 0 and attachment["NetworkCardIndex"] == 0:
+                    private_ip = network_interface["PrivateIpAddress"]
```
Comment (davprat) [inline src/slurm_plugin/fleet_manager.py:54]: Any reason not to break out of this loop if we find what we are looking for?

Comment (gmarciani) [reply src/slurm_plugin/fleet_manager.py:54]: No reason, we should break

**Comment by gmarciani** [comment]
Closed in favour of https://github.com/aws/aws-parallelcluster-node/pull/512

**Pull request closed without merging**